onnxruntime/test/optimizer/qdq_test_utils.h

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "graph_transform_test_builder.h"
+
+#include "core/optimizer/qdq_transformer/selectors_actions/qdq_selector_action_transformer.h"
+#include "core/session/inference_session.h"
+
+#include "test/util/include/asserts.h"
+#include "test/util/include/inference_session_wrapper.h"
+
+namespace onnxruntime {
+namespace test {
+
+using GetQDQConvTestCaseFn = std::function<void(ModelTestBuilder& builder)>;
+
+template <typename T>
+typename std::enable_if<IsTypeQuantLinearCompatible<T>::value, NodeArg*>::type
+AddQDQNodePair(ModelTestBuilder& builder, NodeArg* q_input, float scale, T zp = T()) {
+  auto* q_output = builder.MakeIntermediate();
+  auto* dq_output = builder.MakeIntermediate();
+  builder.AddQuantizeLinearNode<T>(q_input, scale, zp, q_output);
+  builder.AddDequantizeLinearNode<T>(q_output, scale, zp, dq_output);
+  return dq_output;
+}
+
+// TODO: for now it just builds a conv qdq graph.
+// can be modified and made it shared among different qdq test graphs associated with other operators
+template <typename InputType, typename WeightType, typename BiasType, typename OutputType>
+GetQDQConvTestCaseFn BuildQDQConvTestCase(const std::vector<int64_t>& input_shape, const std::vector<int64_t>& weights_shape) {
+  return [input_shape, weights_shape](ModelTestBuilder& builder) {
+    auto* input_arg = builder.MakeInput<float>(input_shape, -1.f, 1.f);
+    auto* output_arg = builder.MakeOutput();
+
+    using InputLimits = std::numeric_limits<InputType>;
+    using WeightLimits = std::numeric_limits<WeightType>;
+    using OutputLimits = std::numeric_limits<OutputType>;
+
+    InputType input_min_value = InputLimits::min();
+    InputType input_max_value = InputLimits::max();
+
+    WeightType weight_min_value = WeightLimits::min();
+    WeightType weight_max_value = WeightLimits::max();
+
+    // initializer values are given as random here (i.e. /=2 is arbitrary.)
+    if (std::is_same<WeightType, int8_t>::value) {
```

Comment:
can it be `if constexpr`?
also I don't quite understand the comment - what is random here?

Reply:
thanks for picking up this. confirmed again that /=2 is not arbitrary it comes from that mlas kernel will overflow on cpu with avx512 extension. the test cases are from the qdq transformer tests. so originally reduced weight range can prevent that overflow.

added comments for this.

Reply:
and what I meant random here is the `MakeInitializer` method below of modeltestbuilder actually uses random value for data

Reply:
ok, thanks for clarifying the comment